### ISSUE TITLE
Formally verify the kernel

### ISSUE THREAD

**Issue by Ticki** [redox-os/redox#521]
+assign @me

**Comment by Roxxik** [comment]
Wouldn't you need to prove some formal property of rustc first?

**Comment by jackpot51** [comment]
The proofs can be seperated, so that you could have something like:

Assuming that rustc correctly compiles this code, the following would be true...

Then, eventually, to have a real proof, you would need to prove rustc.

**Comment by Ticki** [comment]
No, that's not how it works.

First and foremost, rustc can't reason about unsafe code, we need manual reasoning about this. Secondly, the design of the kernel is something that should be verified too.

**Comment by logannc** [comment]
IIRC, L4 verified the spec of their kernel first, then later proved that
the implementation matched the spec. That seems like the way to go.

**Comment by Ticki** [comment]
@logannc That is correct.

**Comment by jackpot51** [comment]
I am not wrong. From the seL4 wiki:

The NICTA team also proved correctness of the translation from C to executable machine code, taking the compiler out of the trusted computing base of seL4. This implies that the high-level security proofs hold for the kernel executable.

**Comment by Ticki** [comment]
Problem is, that Rust is a much more complex language then C.

**Comment by jackpot51** [comment]
Without a proof of whether or not rustc will emit the correct binary code, the proof looses meaning when compilation happens

**Comment by Ticki** [comment]
> Without a proof of whether or not rustc will emit the correct binary code, the proof looses meaning when compilation happens

No, it does not. We're interested in the correctness of Redox, not Rust. We simply have to take certain assumptions, namely that Rust's statical analysis is sound.

**Comment by Roxxik** [comment]
"assuming the machine does what we think...", and your proof gone poof ;) at least for a non opensourced hardware design

**Comment by leeoniya** [comment]
this rabbit hole is bottomless. the physical circuit would have to be formally verfied down to each gate and transistor. any errata or mfg defect and "poof".

at the end of the day you have to draw the line somewhere sane. i think relying on rust's guarantees initially is perfectly fine until rustc itself is formally verified. sure, until that happens the Redox's kernel verification will be weaker that seL4.

is it possible to bridge to gap by transpiling rust's IR to a language that already has tooling to automate some of the verification?

**Comment by jackpot51** [comment]
I'm with @Ticki, make no mistake, we should try to verify the kernel with whatever assumptions we need, then reduce the number of assumptions over time (i.e. compiler, hardware, etc)

**Comment by nixpulvis** [comment]
All this being said, it's going to still be hard because, IIRC rust doesn't even have all the formal guarantees it needs (yet) see topics like https://github.com/rust-lang/rfcs/issues/1447.

**Comment by LifeIsStrange** [comment]
It seems you are lucky :)
https://news.ycombinator.com/item?id=16302530
